modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RerouteProcessor.java

```diff
@@ -94,10 +94,15 @@ public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
         ingestDocument.setFieldValue(DATA_STREAM_TYPE, type);
         ingestDocument.setFieldValue(DATA_STREAM_DATASET, dataset);
         ingestDocument.setFieldValue(DATA_STREAM_NAMESPACE, namespace);
-        if (ingestDocument.hasField(EVENT_DATASET)) {
+        // remove fields with dotted field names
+        ingestDocument.getCtxMap().remove(DATA_STREAM_TYPE);
+        ingestDocument.getCtxMap().remove(DATA_STREAM_DATASET);
+        ingestDocument.getCtxMap().remove(DATA_STREAM_NAMESPACE);
```

Comment:
To me, it seems to be within the rights of this processor to ensure the invariant that there are no conflicting nested and dotted field notations for the `data_stream.*` fields. I don't see why we need to rely on a generic solution for all processors for this. It is a very specific processor with specific variants. The fields that are written to are not configurable with a field reference such as the ones used in the `set` processor, for example.